modules/types-registry/openspec/changes/add-types-registry-phase1/design.md

```diff
@@ -0,0 +1,208 @@
+- Fine-grained locking for writes
+- Already used in HyperSpot (modkit patterns)
+- Simple API, no async complexity
+
+**Alternatives considered**:
+- `RwLock<HashMap>` → Coarse-grained locking, potential contention
+- `tokio::sync::RwLock` → Async overhead for simple operations
+- SQLite in-memory → Overkill for Phase 1.1
+
+### 4. GTS Entity Model
+
+**Decision**: Define a unified `GtsEntity` model:
+
+```rust
+pub struct GtsEntity {
+    pub id: Uuid,              // Deterministic UUID from GTS ID (OP#5)
+    pub gts_id: String,        // Full GTS identifier string
+    pub kind: GtsEntityKind,   // Type or Instance
+    pub content: serde_json::Value,  // JSON Schema or JSON object
+    pub metadata: GtsEntityMetadata,
+    pub created_at: DateTime<Utc>,
+    pub updated_at: DateTime<Utc>,
+}
+
+pub enum GtsEntityKind {
+    Type,      // Schema definition (ends with ~)
+    Instance,  // Object instance (no ~ suffix)
+}
+
+pub struct GtsEntityMetadata {
+    pub description: Option<String>,
+    pub tags: Vec<String>,
+    pub vendor: String,
+    pub package: String,
+    pub namespace: String,
+    pub type_name: String,
+    pub version_major: u32,
+    pub version_minor: Option<u32>,
+}
+```
+
+**Rationale**:
+- Captures all GTS identifier components for efficient querying
+- Stores content as `serde_json::Value` for flexibility
+- Metadata extracted from GTS ID for filtering without parsing
+
+### 5. API Method Signatures
+
+**Decision**: All SDK API methods accept `&SecurityCtx` as first parameter:
+
+```rust
+#[async_trait]
+pub trait TypesRegistryApi: Send + Sync {
+    async fn register(
+        &self,
+        ctx: &SecurityCtx,
+        entity: NewGtsEntity,
+    ) -> Result<GtsEntity, TypesRegistryError>;
+
+    async fn list(
+        &self,
+        ctx: &SecurityCtx,
+        query: ListQuery,
+    ) -> Result<Page<GtsEntity>, TypesRegistryError>;
+
+    async fn get(
+        &self,
+        ctx: &SecurityCtx,
+        gts_id: &str,
+    ) -> Result<GtsEntity, TypesRegistryError>;
+}
+```
+
+**Rationale**:
+- Consistent with HyperSpot conventions
+- Enables future tenant isolation (Phase 1.3)
+- Supports audit logging
+
+### 6. Optional Validation on Registration
+
+**Decision**: Support optional validation flag on registration:
+
+```rust
+pub struct NewGtsEntity {
+    pub gts_id: String,
+    pub content: serde_json::Value,
+    pub description: Option<String>,
+    pub tags: Vec<String>,
+    pub validate_references: bool,  // If true, validate x-gts-ref references
+}
+```
+
+**Rationale**:
+- Allows fast registration without validation for trusted sources
+- Enables strict validation for untrusted input
+- Matches gts-rust registry behavior
+
+## Data Model
+
+### Storage Structure
+
+```
+InMemoryStorage {
+    entities: DashMap<String, GtsEntity>,  // Key: GTS ID string
+    by_uuid: DashMap<Uuid, String>,        // UUID → GTS ID lookup
+    by_vendor: DashMap<String, HashSet<String>>,  // Vendor → GTS IDs index
```

Comment:
Dashmap is thread-safe but HashSet is not thread-safe. Need to double-check if it's ok for us.